demos/induction_heads_multigpu.py

```diff
@@ -0,0 +1,285 @@
+            FlexModel
+        module_names: A list of the module names to which we have attached
+            hooks
+        sequence_length: The sequence length of the prompt passed into the
+            model
+    """
+    induction_score_store = torch.zeros(
+        (
+            config.num_hidden_layers,
+            config.num_attention_heads,
+        ),
+        device=torch.cuda.current_device(),
+    )
+
+    for i, module_name in enumerate(module_names):
+        attn_maps = activation_dict[module_name][0].detach().to(
+            torch.cuda.current_device(),
+        )
+        induction_stripe = attn_maps.diagonal(
+            dim1=-2, dim2=-1, offset=1-sequence_length,
+        )
+        induction_score = einops.reduce(
+            induction_stripe, "batch head_index position -> head_index", "mean",
+        )
+        induction_score_store[i, :] = induction_score
+
+    plt.imshow(induction_score_store.cpu().numpy(), origin="lower")
+    plt.xlabel("Head")
+    plt.ylabel("Layer")
+    plt.title("Induction Score by Head")
+    plt.colorbar()
+    plt.savefig("induction_score_by_head.png", bbox_inches="tight")
+
+
+def get_module_names(config: LlamaConfig) -> list[str]:
+    """Return the list of module names to apply hooks onto.
+
+    Args:
+    ----
+        config: The model's config
+
+    Returns:
+    -------
+        A list of model names that we're applying HookFunctions to
+    """
+    name_placeholder = """_fsdp_wrapped_module.model.layers.
+        {}._fsdp_wrapped_module.self_attn.dummy"""
+    module_names = [
+        name_placeholder.format(i) for i in range(config.num_hidden_layers)
+    ]
+    return module_names
+
+
+def calculate_per_token_loss(
+        logits: torch.Tensor,
+        prompt: torch.Tensor,
+    ) -> None:
+    """Calculate and plot the cross-entropy loss per token.
+
+    Args:
+    ----
+        logits: The model's output logits
+        prompt: The input prompt sequence
+    """
+    # Calculate per token loss
+    log_probs = F.log_softmax(logits, dim=-1)
+    predicted_log_probs = -log_probs[..., :-1, :].gather(
+        dim=-1, index=prompt[..., 1:, None],
+    )[..., 0]
+
+    # Average loss across the batch dimension
+    loss_by_position = einops.reduce(
+        predicted_log_probs, "batch position -> position", "mean",
+    )
+
+    plt.plot(
+        list(range(len(loss_by_position))),
+        loss_by_position.detach().cpu().numpy(),
+    )
+    plt.xlabel("Token Index")
+    plt.ylabel("Loss")
+    plt.title("Loss by position on random repeated tokens")
+    plt.savefig("induction_loss.png", bbox_inches="tight")
+
+
+def main(args: Namespace) -> None:
+    """Execute main demo.
+
+    Args:
+    ----
+        args: Command-line arguments
+    """
+    rank = int(os.environ.get("LOCAL_RANK", 0))
+    torch.cuda.set_device(rank)
+
+    seq_len = 50
+    prompt = torch.randint(
+        500, 15000, (4, seq_len)).to(torch.cuda.current_device(),
+    )
```

Comment:
Maybe alias these numbers to make it apparent what they represent. May also be interesting to make them configurable from the command line